### ISSUE TITLE
Allow scrolloff for global-local

### ISSUE THREAD

**Issue by markwu** [neovim/neovim#11006]
### My environment

* neovim: v0.40-dev, HEAD-11fe132
* macvim: 8.1, patches: 1-1972

### Actual behaviour
`scrolloff` for global only in neovim, but for both `global-local` in recent vim.

### Expected behaviour

I can use the following code in vim.
``` vim
autocmd FileType help setlocal scrolloff=999
```

But, I have to emulate the same behavior by the following code in neovim.
```vim
autocmd BufEnter * if &ft=='help' | let orig_scroll_off = &scrolloff | let &scrolloff = 999 | endif
autocmd BufLeave * if &ft=='help' | let &scrolloff = orig_scroll_off | endif
```

### Reference
[Make 'scrolloff' and 'sidescrolloff' window global-local](https://github.com/vim/vim/pull/3539)

**Comment by markwu** [comment]
It seems [patch 8.1.0864](https://github.com/vim/vim/commit/375e3390078e740d3c83b0c118c50d9a920036c7) done this.

```
patch 8.1.0864: cannot have a local value for 'scrolloff' and 'sidescrolloff'

Problem:    Cannot have a local value for 'scrolloff' and 'sidescrolloff'.
            (Gary Holloway)
Solution:   Make 'scrolloff' and 'sidescrolloff' global-local. (mostly by
            Aron Widforss, closes #3539)
```

**Comment by blueyed** [comment]
@markwu 
Thanks for the reference(s).
Do you want to give it a try?
(`scripts/vim-patch.sh -P 8.1.0864` from Neovim's source should get you started)

**Comment by markwu** [comment]
Here comes the result after I execute `scripts/vim-patch.sh -P 8.1.0864`:

```
$ ~/Temp/neovim$ ./scripts/vim-patch.sh -P 8.1.0864
Cloning Vim into: /Users/mark/Temp/neovim/.vim-src
Cloning into '/Users/mark/Temp/neovim/.vim-src'...
remote: Enumerating objects: 65, done.
remote: Counting objects: 100% (65/65), done.
remote: Compressing objects: 100% (36/36), done.
remote: Total 100995 (delta 37), reused 42 (delta 29), pack-reused 100930
Receiving objects: 100% (100995/100995), 106.72 MiB | 373.00 KiB/s, done.
Resolving deltas: 100% (84861/84861), done.
✔ Found Vim revision '375e3390078e740d3c83b0c118c50d9a920036c7'.
Creating patch...
Pre-processing patch...
✔ Saved patch to '/Users/mark/Temp/neovim/vim-8.1.0864.patch'.

Fetching "origin/master".
✔ From https://github.com/neovim/neovim
 * branch                master     -> FETCH_HEAD

Creating new branch 'vim-8.1.0864' based on 'origin/master'.
✔ Switched to a new branch 'vim-8.1.0864'
Branch 'vim-8.1.0864' set up to track remote branch 'master' from 'origin'.

Creating empty commit with correct commit message.
✔ [vim-8.1.0864 e450424bf] vim-patch:8.1.0864: cannot have a local value for 'scrolloff' and 'sidescrolloff'

Applying patch...
patching file runtime/doc/options.txt
Hunk #1 FAILED at 6588.
Hunk #2 FAILED at 7154.
2 out of 2 hunks FAILED -- saving rejects to file runtime/doc/options.txt.rej
patching file src/nvim/edit.c
Hunk #1 FAILED at 726.
1 out of 1 hunk FAILED -- saving rejects to file src/nvim/edit.c.rej
patching file src/nvim/ex_cmds.c
Hunk #1 FAILED at 3782.
Hunk #2 FAILED at 4388.
2 out of 2 hunks FAILED -- saving rejects to file src/nvim/ex_cmds.c.rej
patching file src/nvim/ex_docmd.c
Hunk #1 FAILED at 8921.
1 out of 1 hunk FAILED -- saving rejects to file src/nvim/ex_docmd.c.rej
can't find file to patch at input line 143
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/src/nvim/misc2.c b/src/nvim/misc2.c
|index 657e16491..278cc330c 100644
|--- a/src/nvim/misc2.c
|+++ b/src/nvim/misc2.c
--------------------------
File to patch:
Skip this patch? [y]
Skipping patch.
1 out of 1 hunk ignored
patching file src/nvim/move.c
Hunk #1 FAILED at 190.
Hunk #2 FAILED at 213.
Hunk #3 FAILED at 267.
Hunk #4 FAILED at 324.
Hunk #5 FAILED at 353.
Hunk #6 FAILED at 371.
Hunk #7 FAILED at 387.
Hunk #8 FAILED at 420.
Hunk #9 FAILED at 446.
Hunk #10 FAILED at 462.
Hunk #11 FAILED at 474.
Hunk #12 FAILED at 946.
Hunk #13 FAILED at 1030.
Hunk #14 FAILED at 1081.
Hunk #15 FAILED at 1101.
Hunk #16 FAILED at 1135.
Hunk #17 FAILED at 1467.
Hunk #18 FAILED at 1525.
Hunk #19 FAILED at 1669.
Hunk #20 FAILED at 1845.
Hunk #21 FAILED at 1902.
Hunk #22 FAILED at 1928.
Hunk #23 FAILED at 1974.
Hunk #24 FAILED at 2128.
Hunk #25 FAILED at 2142.
Hunk #26 FAILED at 2267.
Hunk #27 FAILED at 2285.
27 out of 27 hunks FAILED -- saving rejects to file src/nvim/move.c.rej
patching file src/nvim/normal.c
Hunk #1 FAILED at 2812.
Hunk #2 FAILED at 4633.
Hunk #3 FAILED at 4690.
Hunk #4 FAILED at 4873.
Hunk #5 FAILED at 4895.
5 out of 5 hunks FAILED -- saving rejects to file src/nvim/normal.c.rej
patching file src/nvim/option.c
Hunk #1 FAILED at 225.
Hunk #2 FAILED at 2333.
Hunk #3 FAILED at 2490.
Hunk #4 FAILED at 3736.
Hunk #5 FAILED at 9390.
Hunk #6 FAILED at 10665.
Hunk #7 FAILED at 10759.
Hunk #8 FAILED at 10819.
Hunk #9 FAILED at 13118.
9 out of 9 hunks FAILED -- saving rejects to file src/nvim/option.c.rej
patching file src/nvim/option.h
Hunk #1 FAILED at 1150.
1 out of 1 hunk FAILED -- saving rejects to file src/nvim/option.h.rej
patching file src/nvim/search.c
Hunk #1 FAILED at 2599.
Hunk #2 FAILED at 2635.
Hunk #3 FAILED at 2651.
Hunk #4 FAILED at 2672.
4 out of 4 hunks FAILED -- saving rejects to file src/nvim/search.c.rej
can't find file to patch at input line 913
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/src/nvim/structs.h b/src/nvim/structs.h
|index ddc56bca4..e6cc8291f 100644
|--- a/src/nvim/structs.h
|+++ b/src/nvim/structs.h
--------------------------
File to patch:
Skip this patch? [y] y
Skipping patch.
1 out of 1 hunk ignored
patching file src/nvim/testdir/test_options.vim
Hunk #1 succeeded at 488 (offset 7 lines).
patching file src/nvim/window.c
Hunk #1 FAILED at 4592.
Hunk #2 FAILED at 5873.
2 out of 2 hunks FAILED -- saving rejects to file src/nvim/window.c.rej

Instructions:
  Proceed to port the patch.

  Stage your changes ("git add ..."), then use "git commit --amend" to commit.

  To port more patches (if any) related to 8.1.0864,
  run "vim-patch.sh" again.
    * Do this only for _related_ patches (otherwise it increases the
      size of the pull request, making it harder to review)

  When you are done, try "vim-patch.sh -s" to create the pull request.

  See the wiki for more information:
    * https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-vim
```

**Comment by blueyed** [comment]
@markwu 
This is expected - it's not an easy patch.
You could start from looking at the `.rej` files and apply the patches manually from there.

No worries if you do not want to do it though - it would just be a good way of fixing your issue.. :)

**Comment by blueyed** [comment]
For src/nvim/misc2.c and src/nvim/structs.h you would need to find the current location in Neovim's source.
But other patches might be required / should be ported first maybe.

In general you can use `scripts/vim-patch.sh -L path_to_file` to list missing patches for a file.

**Comment by markwu** [comment]
Ha ... I will leave this issue to who knows how to solve it, it's a little bit difficult to me. Thanks for your information.

**Comment by WillEccles** [comment]
I just wrote a patch in nvim source that fixes this because I did not do my homework and did not see this issue. I set `scrolloff` to be local to the window and added `wo_so` to `winopt_T`. Quick replacement of `p_so` throughout source with `curwin->w_p_so` seems to do the trick. Need to do a little cleanup. Is it acceptable to update in this manner, or does this need to be pulled from vim?

**Comment by stsewd** [comment]
This is already in Neovim, so this can be closed. Although, I find it confusing that it's per window rather than per buffer, but that's another problem.

**Comment by WillEccles** [comment]
> Although, I find it confusing that it's per window rather than per buffer, but that's another problem.

I think that came up in the discussion in #11854 at some point if you want to take a look, but I could be wrong.

**Comment by markwu** [comment]
Thanks @WillEccles , I will close this issue.

**Comment by adigitoleo** [comment]
@willeccles Thanks for working on this. Maybe I'm doing it wrong, but
 `autocmd TermOpen * setlocal scrolloff=0` 
seems to override my global value of `set scrolloff=3` for all windows. Is that the wrong way to use this?

**Comment by willeccles** [comment]
> @willeccles Thanks for working on this. Maybe I'm doing it wrong, but
>  `autocmd TermOpen * setlocal scrolloff=0` 
> seems to override my global value of `set scrolloff=3` for all windows. Is that the wrong way to use this?

No, that should be the right way. That's what I have in my config and it works like a charm. Is it possible your nvim version doesn't have this? If you do `:h scrolloff` it should say it's global-local if yours has the patch.

**Comment by stsewd** [comment]
@adigitoleo keep in mind that scrolloff is local to a window, not to a buffer

**Comment by adigitoleo** [comment]
@willeccles  Oops, you're right I don't have that feature. `:h scrollof` just says "global". I'll wait for the patch. I just found [vim-floaterm](https://github.com/voldikss/vim-floaterm) which is replacing builting `:terminal` for me anyway :)

@stsewd So does that mean if I used those autocommands, and opened a `:terminal`, and then did `<C-/><C-n>` `:b [other buffer]` my global scrolloff setting would not come back?
